pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -1547,29 +1561,35 @@ func (qjm *XController) addOrUpdateCondition(aw *arbv1.AppWrapper, condType arbv
 }
 
 func (qjm *XController) backoff(q *arbv1.AppWrapper, reason string, message string) {
-	var workingAW *arbv1.AppWrapper
-	apiCacheAWJob, e := qjm.queueJobLister.AppWrappers(q.Namespace).Get(q.Name)
-	// Update condition
-	if e == nil {
-		workingAW = apiCacheAWJob
-		apiCacheAWJob.Status.QueueJobState = arbv1.AppWrapperCondBackoff
-		workingAW.Status.FilterIgnore = true // update QueueJobState only, no work needed
-		qjm.addOrUpdateCondition(workingAW, arbv1.AppWrapperCondBackoff, v1.ConditionTrue, reason, message)
-		//qjm.updateEtcd(workingAW, "backoff - Rejoining")
-		qjm.updateStatusInEtcd(workingAW, "backoff - Rejoining")
-	} else {
-		workingAW = q
-		klog.Errorf("[backoff] Failed to retrieve cached object for %s/%s.  Continuing with possible stale object without updating conditions.", workingAW.Namespace, workingAW.Name)
 
+	etcUpdateRetrier := retrier.New(retrier.ExponentialBackoff(10, 100*time.Millisecond), &EtcdErrorClassifier{})
```

Comment:
backoff is another thread while backoff retries can scheduleNext thread pick same AW and start working?

Reply:
Yes, backoff is called if Schedule next can't dispatch an AppWrappers. It runs asynchronously.

Reply:
So, post-fix there is not race condition?

Reply:
There are still race conditions in the updates to etc, retry allows you to deal with them.